File: android/app/src/main/java/com/driveincar/domain/race/SubmitTimeUseCase.kt
package com.driveincar.domain.race

import com.driveincar.data.ranking.RankingRepository
import com.driveincar.data.user.UserRepository
import javax.inject.Inject

class SubmitTimeUseCase @Inject constructor(
    private val rankingRepo: RankingRepository,
    private val userRepo: UserRepository,
) {
    suspend operator fun invoke(
        uid: String,
        courseId: String,
        timeMs: Long,
        averageKmh: Double,
        flagged: Boolean,
    ): Result<String> {
        val user = userRepo.fetchUser(uid)
            ?: return Result.failure(IllegalStateException("user not found: $uid"))

        return rankingRepo.submitRanking(
            courseId = courseId,
            uid = uid,
            nickname = user.nickname,
            carDisplay = user.carDisplay,
            profileImageId = user.profileImageId,
            timeMs = timeMs,
            averageKmh = averageKmh,
            flagged = flagged,
        )
    }
}

File: ios/DriveInCar/Domain/Race/SubmitTimeUseCase.swift
import Foundation

struct SubmitTimeUseCase {
    let users: UserRepository
    let rankings: RankingRepository

    func callAsFunction(
        uid: String,
        courseId: String,
        timeMs: Int64,
        averageKmh: Double,
        flagged: Bool
    ) async throws -> String {
        guard let user = try await users.fetchUser(uid: uid) else {
            throw NSError(
                domain: "DriveInCar.SubmitTime",
                code: 404,
                userInfo: [NSLocalizedDescriptionKey: "user not found: \(uid)"]
            )
        }
        return try await rankings.submit(
            courseId: courseId,
            uid: uid,
            nickname: user.nickname,
            carDisplay: user.carDisplay,
            profileImageId: user.profileImageId,
            timeMs: timeMs,
            averageKmh: averageKmh,
            flagged: flagged
        )
    }
}
